### PR TITLE
CDI-658 Introduce ProducerConfigurator

### PR DESCRIPTION
**Pull request by mkouba**

### CODE REVIEW

File: api/src/main/java/javax/enterprise/inject/spi/ProcessProducer.java [hunk 1]
```diff
@@ -73,6 +77,21 @@
...
+     * <p>
+     * Each call returns the same configurator instance within an observer notification.
+     * </p>
+     * 
+     * @return a non reusable {@link ProducerConfigurator} to configure the original
+     *         {@link javax.enterprise.inject.spi.Producer}.
+     * @throws IllegalStateException if called outside of the observer method invocation
+     * @since 2.0
+     */
+    public ProducerConfigurator<X> configureProducer();
```
Comment (antoinesd) [inline api/src/main/java/javax/enterprise/inject/spi/ProcessProducer.java]: for consistency with other Lifecycle events where we provide access to configurators, thus method should be renamed to `setProducer()`

File: api/src/main/java/javax/enterprise/inject/spi/configurator/ProducerConfigurator.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    <U extends T> ProducerConfigurator<T> produceWith(Function<CreationalContext<U>, U> callback);
+
+    /**
+     * Set a callback to destroy the produced instance.
+     * 
+     * @param callback
+     * @return self
+     * @see Producer#dispose(Object)
+     */
+    ProducerConfigurator<T> disposeWith(Consumer<T> callback);
```
Comment (antoinesd) [inline api/src/main/java/javax/enterprise/inject/spi/configurator/ProducerConfigurator.java]: why not. But why only here and not `produceWith()` as well?

File: api/src/main/java/javax/enterprise/inject/spi/ProcessProducer.java (hunk 1 above)
Comment (mkouba) [inline api/src/main/java/javax/enterprise/inject/spi/ProcessProducer.java]: `ProcessAnnotatedType.configureAnnotatedType()`, `ProcessBeanAttributes.configureBeanAttributes()` ...?

File: api/src/main/java/javax/enterprise/inject/spi/configurator/ProducerConfigurator.java (hunk 2 above)
Comment (mkouba) [inline api/src/main/java/javax/enterprise/inject/spi/configurator/ProducerConfigurator.java]: I would leave it as it is for 2.0.

File: api/src/main/java/javax/enterprise/inject/spi/ProcessProducer.java (hunk 1 above)
Comment (antoinesd) [inline api/src/main/java/javax/enterprise/inject/spi/ProcessProducer.java]: forget this, I'm lacking sleep...

**Comment by manovotn** [comment]
Looks good

**Pull request merged**